cheatsheets/gleam-for-typescript-users.md

```diff
@@ -0,0 +1,794 @@
+{a: 1} == {a: 1} // false
+```
+
+Maps in TypeScript are a class that are similar to objects, but separate.
+
+```ts
+const myMap = new Map<string, number>;
+
+myMap.set('a', 1);
+myMap.has('a'); // true
+myMap.get('a'); // 1
+```
+
+#### Gleam
+
+A dict, in Gleam, is a collection of keys and values which other languages may call a hashmap or table.
+
+```gleam
+import gleam/io
+import gleam/dict
+
+pub fn main() {
+  let scores = dict.from_list([#("Lucy", 13), #("Drew", 15)])
+  io.debug(scores) // dict.from_list([#("Drew", 15), #("Lucy", 13)])
+
+  let scores =
+    scores
+    |> dict.insert("Bushra", 16)
+    |> dict.insert("Darius", 14)
+    |> dict.delete("Drew")
+  io.debug(scores) // dict.from_list([#("Darius", 14), #("Bushra", 16), #("Lucy", 13)])
+}
+```
+
+### Sets
+
+#### TypeScript
+
+```ts
+const nums = new Set();
+
+nums.add(1);
+nums.has(1); // true
+```
+
+#### Gleam
+
+```gleam
+import gleam/set
+
+set.new()
+|> set.insert(2)
+|> set.contains(2)
+// -> True
+```
+## Functions
+
+Both TypeScript and Gleam have first-class functions. Both support functions as values,
+as well as higher-order functions, partial application, currying, and similar patterns.
+
+### TypeScript
+
+```ts
+// Traditional function declaration
+function example4() {
+  return 'a';
+}
+
+// Declaration as a variable with an anonymous function
+const example2 = function() {
+  return 'b';
+}
+
+// Anonymous arrow function expressions
+const example3 = () => 'c'; // implicit return
+const addOne = x => x + 1; // single arguments don't require parentheses
+
+const example4 = example3; // functions can be passed as values
+
+example4(); // 'c'
+```
+
+### Gleam
+
+The basic syntax for functions in Gleam is similar. One key difference is that the last
+expression in a function is returned implicitly without a `return` keyword.
+
+```gleam
+fn identity(x) {
+  x
+}
+
+fn main() {
+  let func = identity
+  func(100) // 100
+
+  let func2 = fn() { "b" }
+
+  func2() // "b"
+}
+```
+
+## Blocks
+
+### Gleam
+
+Gleam has support for blocks, which are similar to functions without arguments.
+
+```gleam
+let example = {
+  let ratio = get_ratio()
+  some_value * ratio()
+}
+
+// Also used for explicit math precendence.
+let total = { x - 3} * 12
+```
+
+### TypeScript
+
+The equivalent in TypeScript would be an [IIFE (Immediately Invoked Function Expression)](https://developer.mozilla.org/en-US/docs/Glossary/IIFE).
+
+```ts
+const example = (() => {
+  const ratio = getRatio();
+
+  return someValue * ratio;
+})();
+```
+
+## Flow Control
+
+### Conditionals
+
+#### TypeScript
+
+```ts
+if (x === 1) {
+  console.log('one');
+} else if (x === 2) {
+  console.log('two');
+} else {
+  console.log('something else');
+}
+```
+
+#### Gleam
+
+Gleam does not have the same if/else syntax and uses `case` statements/expressions instead.
+
+if/else example:
+
+```gleam
+case is_admin {
+  True -> "allow access"
+  False -> "disallow access"
+}
+```
+
+if/elseif/else example:
```

Comment:
the first example is good, but i'm not sure if the second example is. perhaps a good way to showcase a more complex example of 'case' would be to do some nesting inside the True arm? eg. another case statement

```import gleam/io

pub fn main() {
  let x = 3
  let y = 2

  let result = case x {
    // Match specific values
    0 -> {
      case y {
        1 -> "hello"
        _ -> "good byee"
      }
    }
    1 -> "One"

    // Match any other value
    _ -> "Other"
  }
  io.debug(result)
}